Introducing table sayall commands

mltony opened the pull request:
### Link to issue number:
Fixes #13469
### Summary of the issue:
Feature request: add table sayAll commands to read rows and columns.
### Description of how this pull request fixes the issue:
Adds 4 commands:
* SayAll from current cell horizontally to the right.
* SayAll from current cell vertically down.
* Speak current column from top to bottom without moving caret.
* speak current row from left to right without moving caret.

Additionally:
* Refactored class _TextReader in sayAll.py:
    * Refactored out a few methods for different types of cursor and made them abstract.
    * Added implementations for review and caret cursor types and also for new table cursor type.
* Refactored class DocumentWithTableNavigation:
    * Refactored cell moving logic as method _tableFindNewCell, which is used in both simple table commands and sayAll commands.
    * Added _TableCell class to replace 5-element tuples that contain information about table cells.
### Testing strategy:
* Tested sayall commands in Chrome, Firefox, MSWord (both browse and focus mode); tested with merged cels to check that they are handled correctly during sayAll.
* Tested normal table navigation commands to avoid regression.
* Tested sayAll commands in Chrome, Firefox, MSWord to avoid regressions.

### Known issues with pull request:
* Doesn't support Excel tables.
* Doesn't support table-like list views.
### Change log entries:
New features
* Added table sayAll commands.
Changes
Bug fixes
For Developers

### Code Review Checklist:
- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

mltony committed:
Refactoring sayAll commands

mltony committed:
Refactoring table navigation commands

mltony committed:
Adding table sayAll commands

mltony committed:
Update documentation

mltony committed:
Storing true row/col before invoking table sayAll in order to handle merged cells correctly during sayAll

mltony commented:
I see FAIL: Translation comments check. Not quite sure how to investigate that. It suggests to find more information in the build log, however among 5k lines of log all unit tests seem to succeed and no obvious translation-related errors can be found.

josephsl commented:
Hi, comments for translators are needed for any string meant to be shown to users or spoken/brailled. This includes input help messages for new scripts. Thanks.

mltony committed:
Trying to please Translation comments check

mltony commented:
Fixed translation check.

cary-rowen commented:
Great job, I tested it and it seems to be performing well so far.

XLTechie commented on source/documentBase.py:
```diff
@@ -48,6 +49,15 @@ class _TableSelection:
 
 
+@dataclass
+class _TableCell:
```
Shouldn't this have a docstring?

lukaszgo1 commented on source/documentBase.py:
```diff
@@ -100,13 +110,14 @@ def _maybeGetLayoutTableIds(self, info: textInfos.TextInfo):
+	def _getTableCellCoords(
+			self,
+			info: textInfos.TextInfo,
+	) -> _TableCell:
```
I'm slightly worried about changing return type of this method. While it is private and theoretically should not be called by anything outside `documentBase` this is clearly not the case and some add-on may rely on its result being a tuple and use it as an iterable or rely on the fact that it supports indexing operator. Have you considered using `typing.NamedTuple` - that should give you both type hints and ability to use attribute names rather than rely on indexes, and at the same time preserves backwards compatibility?

feerrenrut replied:
A line has to be drawn somewhere. A Dataclass is much clearer, and understood by code quality tools better than a NamedTuple. Add-ons shouldn't be relying on underscore methods or classes. If required for an add-on, and issue should be opened.

CyrilleB79 commented:
I have just tested the snapshot. It's very nice and works well! Thanks.

One additional point to clarify:
You have not implemented these commands in the list views where table commands can operate. Is it on purpose?
Since this PR already provides fully functional new features, the request can be handled in a subsequent PR if you prefer.
In any case, please indicate if and what you want to do something for list view commands.

feerrenrut commented:
@mltony Please consider adding unit and/or system tests for this behavior.

seanbudd commented:
We won't take this on until there is system tests for this behaviour and unit tests where possible.
We will schedule this work if nobody else picks this up by that time.

mltony commented:
Yes, I will work on that. Sorry for delay. I got overwhelmed by projects at work, but hopefully next month I'll have some time to address all the comments.

mltony committed:
Adding system tests for sayAll commands in chrome

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user guide

mltony committed:
Adding docstring

CyrilleB79 commented on source/documentBase.py:
```diff
@@ -418,6 +506,30 @@ def script_lastColumn(self, gesture):
+		self._tableSayAll(_Movement.NEXT, _Axis.COLUMN)
+	script_sayAllRow.__doc__ = _(
+		# Translators: the description for the sayAll row command
+		"Reads row horizontally from current table cell up untill the last cell in the row."
```
I have made suggestion for the script description in the user guide (see below). If you accept my suggestions, do not forget to align the script descriptions in this file.

mltony committed:
Fix lint

mltony commented:
I addressed all the comments in this thread. This PR is ready for review again.
@XLTechie, added docstring.
@CyrilleB79, I don't think this applies to all list views. Or do you mean table-like views like with multiple columns? You're right this PR doesn't support that. It also doesn't support Excel tables. Sure, this can be handled in separate PRs. Also I corrected documentation in my latest commit.        @feerrenrut, added system test.
@feerrenrut, @seanbudd: I added a system test for table sayAll commands. Silly bot says system tests are failing, but the only failed test "ARIA details :: Ensure a summary of aria-details is read on comman" is not mine - could it be flaky? My test succeeds.

seanbudd commented on source/documentBase.py:
```diff
@@ -48,6 +49,18 @@ class _TableSelection:
+@dataclass
+class _TableCell:
+	"""
+	Contains information about current cell in the table.
```
I think TableCell is too generic to always be the "current cell"

```suggestion
	Contains information about a cell in the table with matching tableID
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,10 +346,10 @@ When within a table, the following key commands are also available:
-| Speak current column | NVDA+control+alt+UpArrow | Reads current column vertically from top to bottom without moving system caret. |
-| Speak current row | NVDA+control+alt+LeftArrow | Reads current row vertically from left to right without moving system caret. |
+| Say current column from current cell down | NVDA+control+alt+DownArrow | Reads column vertically from current table cell down to the last cell in the column. |
+| Say current row from current cell right | NVDA+control+alt+RightArrow | Reads row horizontally from current table cell up untill the last cell in the row. |
```
I think a short name is better here, with more information in the description.

```suggestion
| Speak row rightwards | ``NVDA+control+alt+rightArrow`` | Reads row horizontally from current table cell rightwards to the last cell in the row. |
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,10 +346,10 @@ When within a table, the following key commands are also available:
-| Say all current row | NVDA+control+alt+RightArrow | Reads row horizontally from current table cell up untill the last cell in the row. |
-| Speak current column | NVDA+control+alt+UpArrow | Reads current column vertically from top to bottom without moving system caret. |
-| Speak current row | NVDA+control+alt+LeftArrow | Reads current row vertically from left to right without moving system caret. |
+| Say current column from current cell down | NVDA+control+alt+DownArrow | Reads column vertically from current table cell down to the last cell in the column. |
```
I think a short name is better here, with more information in the description. I've also fixed the formatting of the key commands.

```suggestion
| Speak column downwards | ``NVDA+control+alt+downArrow`` | Reads column vertically from the current table cell downwards to the last cell in the column. |
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,10 +346,10 @@ When within a table, the following key commands are also available:
+| Say current column from current cell down | NVDA+control+alt+DownArrow | Reads column vertically from current table cell down to the last cell in the column. |
+| Say current row from current cell right | NVDA+control+alt+RightArrow | Reads row horizontally from current table cell up untill the last cell in the row. |
+| Speak entire current column | NVDA+control+alt+UpArrow | Reads current column vertically from top to bottom without moving system caret. |
+| Speak entire current row | NVDA+control+alt+LeftArrow | Reads current row vertically from left to right without moving system caret. |
```
```suggestion
| Speak entire row | ``NVDA+control+alt+leftArrow`` | Reads current row vertically from left to right without moving the system caret. |
```

seanbudd commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,10 +346,10 @@ When within a table, the following key commands are also available:
-| Speak current row | NVDA+control+alt+LeftArrow | Reads current row vertically from left to right without moving system caret. |
+| Say current column from current cell down | NVDA+control+alt+DownArrow | Reads column vertically from current table cell down to the last cell in the column. |
+| Say current row from current cell right | NVDA+control+alt+RightArrow | Reads row horizontally from current table cell up untill the last cell in the row. |
+| Speak entire current column | NVDA+control+alt+UpArrow | Reads current column vertically from top to bottom without moving system caret. |
```
```suggestion
| Speak entire column | ``NVDA+control+alt+upArrow`` | Reads current column vertically from top to bottom without moving the system caret. |
```

seanbudd commented on tests/system/robot/chromeTests.py:
```diff
@@ -1705,6 +1705,147 @@ def test_tableNavigationWithMergedColumns():
+	if True:
+		jumpToA3()
+		# sayAll row
+		actualSpeech = _chrome.getSpeechAfterKey("NVDA+Control+Alt+RightArrow")
```
Key commands are generally lowerCamelCase

```suggestion
		actualSpeech = _chrome.getSpeechAfterKey("NVDA+control+alt+rightArrow")
```

seanbudd commented on tests/system/robot/chromeTests.py:
```diff
@@ -1705,6 +1705,147 @@ def test_tableNavigationWithMergedColumns():
+			"row 4  column 3  through row 5 column 4  C 4 plus D 4 plus  C 5 plus D 5"
+		)
+
+	if True:
```
I suggest splitting these `if` blocks up into separate tests eg `_test_SayAllReadEntireColumn`.

`test_tableSayAll` could call all of these.

seanbudd commented:
>You're right this PR doesn't support that. It also doesn't support Excel tables. Sure, this can be handled in separate PRs.

Can you add this to the known issues section in the PR description

>Silly bot says system tests are failing, but the only failed test "ARIA details :: Ensure a summary of aria-details is read on comman" is not mine - could it be flaky? My test succeeds.

Yes frustratingly we have generic flaky test problem.

CyrilleB79 commented:
My comments on https://github.com/nvaccess/nvda/pull/13670/commits/537bdfcfbd930005df79ca2cfe519321238fcbe0 have been marked as resolved by @seanbudd and @mltony has pushed additional modifications in the user guide. However, my comments do not seem to have been taken into account; I do not know if it was intentional or not. If you do not accept them, I would expect some sort of additional explanation. And at least the typo should be fixed.

You can see the comments in the commit or look at the following points:
* typo: untill -> until
* Use "say all" in the name of the NVDA+control+alt+DownArrow/rightArrow commands, in order to use the same vocabulary as the "say all" command to reflect that the caret is moving while the text is being read. E.g.: "Say all in the current column" or shorter "Say all current column" or "say all in column".
* "from current table cell " -> "from the current cell " (twice): drop "table" which is obvious and add "the" to avoid telegraphic style in the command description.
* "without moving system caret. " -> "without moving the system caret. ": add "the" to avoid telegraphic style in command description.
* Do not forget to update the script's description in `documentBase.py` as per the user guide.

In addition, I would recommend to replace "Speak entire current column" with "Read entire current column", to make something similar as "Read line" in the table just above. On the contrary "Speak" in the names of the commands are used for NVDA+2 (speak current character) and NVDA+3 (speak current word), which are other types of commands.

mltony committed:
Update source/documentBase.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update tests/system/robot/chromeTests.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Making keystrokes camel case

mltony committed:
Splitting system test into three

mltony commented:
@seanbudd, all done!

seanbudd committed:
Merge branch 'master' into table_sayall

seanbudd commented on source/documentBase.py:
```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
+		from scriptHandler import isScriptWaiting
+		from speech import speakTextInfo
+
+		if isScriptWaiting():
+			return
```
Suggested an extra new line here

```suggestion
		if isScriptWaiting():
			return

```

seanbudd commented on source/documentBase.py:
```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
+		formatConfig = config.conf["documentFormatting"].copy()
+		formatConfig["reportTables"] = True
+		try:
+			cell, info, tableSelection = self._tableFindNewCell(
```
`_cell` as it is unused.

```suggestion
			_cell, info, tableSelection = self._tableFindNewCell(
```

seanbudd commented on source/documentBase.py:
```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
+			updateCaret: bool = True,
+	) -> None:
+		try:
+			cell, info, tableSelection = self._tableFindNewCell(
```
```suggestion
			_cell, info, tableSelection = self._tableFindNewCell(
```

seanbudd commented on source/documentBase.py:
```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
+
+		def nextLineFunc(info: textInfos.TextInfo) -> textInfos.TextInfo:
+			try:
+				cell, newInfo, tableSelection = self._tableFindNewCell(
```
```suggestion
				_cell, newInfo, tableSelection = self._tableFindNewCell(
```

seanbudd commented on source/documentBase.py:
```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
+		except LookupError as e:
 			# Translators: The message reported when a user attempts to use a table movement command
-			# when the cursor is not within a table.
+			# Translators: when the cursor is not within a table.
```
This change is not needed.

```suggestion
			# when the cursor is not within a table.
```

seanbudd commented on source/documentBase.py:
```diff
@@ -442,4 +557,8 @@ def script_toggleIncludeLayoutTables(self,gesture):
 		"kb:control+alt+pageDown": "lastRow",
 		"kb:control+alt+Home": "firstColumn",
 		"kb:control+alt+End": "lastColumn",
+		"kb:NVDA+control+alt+RightArrow": "sayAllRow",
+		"kb:NVDA+control+alt+DownArrow": "sayAllColumn",
+		"kb:NVDA+control+alt+LeftArrow": "speakRow",
+		"kb:NVDA+control+alt+UpArrow": "speakColumn",
```
```suggestion
		"kb:NVDA+control+alt+rightArrow": "sayAllRow",
		"kb:NVDA+control+alt+downArrow": "sayAllColumn",
		"kb:NVDA+control+alt+leftArrow": "speakRow",
		"kb:NVDA+control+alt+upArrow": "speakColumn",
```

seanbudd reviewed (commented):
Thanks @mltony - I've done another parse of the full PR. This is nearly ready! just a bunch of minor things.

CyrilleB79 commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,6 +346,10 @@ When within a table, the following key commands are also available:
 | Move to last column | control+alt+end | Moves the system caret to the last column (staying in the same row) |
 | Move to first row | control+alt+pageUp | Moves the system caret to the first row (staying in the same column) |
 | Move to last row | control+alt+pageDown | Moves the system caret to the last row (staying in the same column) |
+| Speak column downwards | ``NVDA+control+alt+downArrow`` | Reads column vertically from the current table cell downwards to the last cell in the column. |
+| Speak row rightwards | ``NVDA+control+alt+rightArrow`` | Reads row horizontally from current table cell rightwards to the last cell in the row. |
```
* Use "say all" in the name of the NVDA+control+alt+DownArrow/rightArrow commands, in order to use the same vocabulary as the "say all" command to reflect that the caret is moving while the text is being read. E.g.: "Say all in the current column" or shorter "Say all current column" or "say all in column".
* "Reads column/row" -> "Read the column/row": added "the" to avoid telegraphic style in the command description.
* "from current table cell " -> "from the current cell " (twice): drop "table" which is obvious and add "the" to avoid telegraphic style in the command description.

```suggestion
| Say all in column | ``NVDA+control+alt+downArrow`` | Reads the column vertically from the current cell downwards to the last cell in the column. |
| Say all in row | ``NVDA+control+alt+rightArrow`` | Reads the row horizontally from the current cell rightwards to the last cell in the row. |
```

CyrilleB79 reviewed (commented):
Hi

Due to previous conflicts leading to outdated comments, I put again my updated review on top of the last commit.
The comments are only in the User Guide. But when the wording is validated, do not forget to update the script's description in documentBase.py as per the user guide. Maybe also the script's name? (as you want)

CyrilleB79 commented on user_docs/en/userGuide.t2t:
```diff
@@ -346,6 +346,10 @@ When within a table, the following key commands are also available:
 | Move to last row | control+alt+pageDown | Moves the system caret to the last row (staying in the same column) |
+| Speak column downwards | ``NVDA+control+alt+downArrow`` | Reads column vertically from the current table cell downwards to the last cell in the column. |
+| Speak row rightwards | ``NVDA+control+alt+rightArrow`` | Reads row horizontally from current table cell rightwards to the last cell in the row. |
+| Speak entire column | ``NVDA+control+alt+upArrow`` | Reads current column vertically from top to bottom without moving the system caret. |
+| Speak entire row | ``NVDA+control+alt+leftArrow`` | Reads current row vertically from left to right without moving the system caret. |
```
Another comment forgotten in the previous review:
* Replace "Speak entire column/row" with "Read entire column/row", to make something similar as "Read current line" and "Read current selection" in the table just above. On the contrary "Speak" in the names of commands are used only for "Speak typed characters/typed words/command keys" which are other types of commands.
* "Read current column/row" -> "Read the current column/row":added "the" to avoid telegraphic style.

```suggestion
| Read entire column | ``NVDA+control+alt+upArrow`` | Reads the current column vertically from top to bottom without moving the system caret. |
| Read entire row | ``NVDA+control+alt+leftArrow`` | Reads the current row vertically from left to right without moving the system caret. |
```

mltony committed:
Update source/virtualBuffers/__init__.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/virtualBuffers/gecko_ia2.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update source/virtualBuffers/gecko_ia2.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony commented on tests/system/robot/chromeTests.py:
```diff
@@ -1744,106 +1740,122 @@ def test_tableSayAll():
+	prepareChromeForTableSayAllTests()
+	tableSayAllJumpToB2()
+	# Speak current column
+	actualSpeech = _chrome.getSpeechAfterKey("NVDA+control+alt+upArrow")
+	_asserts.strings_match(
```
chromeLib didn't have getSpeechAndBrailleAfterKey function. I added one - just forwarded call to nvdaLib. However it appears to return wrong result - perhaps it doesn't work well with sayAll commands. Or perhaps Braille itself doesn't work with sayall as speech? I myself don't use Braille, nor do I have a Braille display to test this.

mltony committed:
Update tests/system/robot/chromeTests.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Update user_docs/en/userGuide.t2t

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

mltony committed:
Addressing comments

mltony committed:
Merge branch 'table_sayall' of github.com:mltony/nvda into table_sayall

mltony committed:
Updating tests

mltony committed:
Updating unit tests

seanbudd commented on tests/system/libraries/ChromeLib.py:
```diff
@@ -194,6 +197,10 @@ def getSpeechAfterKey(key) -> str:
 		return _NvdaLib.getSpeechAfterKey(key)
 
+	@staticmethod
+	def getSpeechAndBrailleAfterKey(key) -> _Tuple[str, str]:
```
Why not just use `_NvdaLib.getSpeechAndBrailleAfterKey` directly?

mltony replied:
Sure, reverted to _NvdaLib.getSpeechAndBrailleAfterKey. But I mean what's the difference - Braille doesn't seem to produce any meaningful output anyway.

seanbudd replied:
Apologies - I was under the assumption that braille had meaningful output for these commands. Feel free to not test braille in that case then.

mltony committed:
Updating tests

seanbudd commented:
Is this ready for review - i.e. no longer a draft?

mltony commented:
Yes, ready for review.

seanbudd commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -23,6 +23,7 @@
 import aria
 import config
 from NVDAObjects.IAccessible import normalizeIA2TextFormatField, IA2TextTextInfo
+import documentBase
```
my mistake, this should not have been introduced. Will push a fix for this.

seanbudd committed:
Merge branch 'master' into table_sayall

seanbudd committed:
fix up missed review comments

seanbudd committed:
fix lint

seanbudd committed:
update changes

seanbudd commented:
Thanks @mltony - I've pushed some minor changes to fix up the review comments missed in `source/documentBase.py`.

Qchristensen reviewed (approved):
User guide changes read well, good work!

seanbudd committed:
fix checkpot
